### PR TITLE
Bug 2057561 - Optimize array mapping and siblings rerenders (Fix 3)

### PR DESCRIPTION
**Pull request by davidmiculit**
[Bug 2057561 - Improve frontend performance on Graphs View](https://bugzilla.mozilla.org/show_bug.cgi?id=2057561)

Toggling the visibility of a single series unnecessarily maps over the entire testData array and triggers expensive re-renders for all sibling LegendCard components.

Fix: 
- Wrap LegendCard in React.memo with a custom comparator on series.signature_id + series.visible.
- Refactor the updateSelectedTest visibility-toggle handler to bypass the full-array .map(). It now performs an O(1) immutable single-item update by finding the target index, cloning the specific item, and rebuilding the array using slice.

### CODE REVIEW

File: ui/perfherder/graphs/LegendCard.jsx [hunk 1]
```diff
@@ -228,4 +233,9 @@ LegendCard.propTypes = {
   selectedDataPoint: PropTypes.shape({}),
 };
 
-export default LegendCard;
+const areEqual = (prev, next) =>
+  prev.series.signature_id === next.series.signature_id &&
+  prev.series.visible === next.series.visible &&
+  prev.colors === next.colors;
+
```
Comment (esanuandra) [inline ui/perfherder/graphs/LegendCard.jsx]:
```suggestion
const areEqual = (prev, next) =>
  prev.series === next.series &&
  prev.testData === next.testData &&
  prev.colors === next.colors &&
  prev.symbols === next.symbols &&
  prev.selectedDataPoint === next.selectedDataPoint;
```

There is a bug with the current implementation
STR:
Select multiple tests to plot on the graph
Uncheck one test from the legend card
Remove the unchecked test
Uncheck another test

It puts back the removed unchecked test

File: ui/perfherder/graphs/LegendCard.jsx [hunk 2]
```diff
@@ -22,42 +22,47 @@ const LegendCard = ({
...
-            "The graph supports viewing 6 tests at a time. To select and view a test that isn't currently visible, first deselect a visible test",
-          );
-        }
-      }
-      return item;
-    });
+    const targetIndex = testData.findIndex((item) => item.signature_id === series.signature_id);
+    const item = testData[targetIndex];
+    const isVisible = !item.visible;
+    let updatedItem = { ...item };
```
Comment (esanuandra) [inline ui/perfherder/graphs/LegendCard.jsx]: Nit: It's not modified in the file

```suggestion
    const updatedItem = { ...item };
```

File: ui/perfherder/graphs/LegendCard.jsx
Comment (esanuandra) [inline ui/perfherder/graphs/LegendCard.jsx:144]: This part of code also needs an update so the identity-based comparison (`prev.series === next.series`, or a value read off it) correctly sees a change.

```suggestion
const promoteIndex = graphColors.length - 1;

if (
newData[promoteIndex] &&
newData[promoteIndex].color[0] === 'border-secondary'
) {
const promoted = newData[promoteIndex];
newData[promoteIndex] = {
      ...promoted,
      color: series.color,
      visible: true,
      data: promoted.data.map((item) => ({ ...item, z: series.color[1] })),
    };
resetParams(newData);
```

File: ui/perfherder/graphs/LegendCard.jsx [hunk 3]
```diff
@@ -9,55 +9,65 @@ import GraphIcon from '../../shared/GraphIcon';
...
+    if (isVisible && newColors.length && newSymbols.length) {
+      updatedItem.color = newColors.pop();
+      updatedItem.symbol = newSymbols.pop();
+      updatedItem.visible = isVisible;
+      updatedItem.data = item.data.map((test) => ({
+        ...test,
+        z: updatedItem.color[1],
+        _z: updatedItem.symbol,
+      }));
+    } else if (!isVisible) {
```
Comment (esanuandra) [inline ui/perfherder/graphs/LegendCard.jsx:44]: Add the queued flag here also

```suggestion
    if (isVisible && newColors.length && newSymbols.length) {
      updatedItem.color = newColors.pop();
      updatedItem.symbol = newSymbols.pop();
      updatedItem.visible = isVisible;
      updatedItem.queued = false;
      updatedItem.data = item.data.map((test) => ({
        ...test,
        z: updatedItem.color[1],
        _z: updatedItem.symbol,
      }));
    } else if (!isVisible) {
```

File: ui/perfherder/graphs/LegendCard.jsx [hunk 4]
```diff
@@ -9,55 +9,65 @@ import GraphIcon from '../../shared/GraphIcon';
...
+      newColors.push(item.color);
+      newSymbols.push(item.symbol);
+      updatedItem.color = ['border-secondary', ''];
+      updatedItem.symbol = ['circle', 'outline'];
+      updatedItem.visible = isVisible;
+      updatedItem.data = item.data.map((test) => ({
```
Comment (esanuandra) [inline ui/perfherder/graphs/LegendCard.jsx:50]: Add the queued flag here also

```suggestion
      updatedItem.color = ['border-secondary', ''];
      updatedItem.symbol = ['circle', 'outline'];
      updatedItem.visible = isVisible;
      updatedItem.queued = false;
      updatedItem.data = item.data.map((test) => ({
```

File: ui/perfherder/graphs/LegendCard.jsx [hunk 5]
```diff
@@ -94,36 +104,50 @@ const LegendCard = ({
...
+      };
       resetParams(newData);
     } else if (series.color[0] === 'border-secondary') {
       resetParams(newData);
     } else {
-      const newColors = [...colors, ...[series.color]];
-      resetParams(newData, newColors);
+      const newColors = [...colors, series.color];
+      const newSymbols = [...symbols, series.symbol];
+      resetParams(newData, newColors, newSymbols);
```
Comment (esanuandra) [inline ui/perfherder/graphs/LegendCard.jsx:150]: This should fix some of the edge cases bugs related to the hidden tests that are removed.

```suggestion
      // Removing a hidden test frees no color slot → just drop it.
    if (series.color[0] === 'border-secondary') {
      resetParams(newData);
      return;
    }

    // A visible test was removed → promote the next auto-queued test (one that
    // was force-hidden only because the palette was exhausted). Tests the user
    // deliberately hid are queued:false and are never auto-promoted.
    const promoteIndex = newData.findIndex((item) => item.queued);

    if (promoteIndex !== -1) {
      const promoted = newData[promoteIndex];
      newData[promoteIndex] = {
        ...promoted,
        color: series.color,
        symbol: series.symbol,
        visible: true,
        queued: false,
        data: promoted.data.map((item) => ({
          ...item,
          z: series.color[1],
          _z: series.symbol,
        })),
      };
      resetParams(newData);
      return;
    }

    // No auto-queued test waiting → return the freed color/symbol to the pool.
    const newColors = [...colors, series.color];
    const newSymbols = [...symbols, series.symbol];
    resetParams(newData, newColors, newSymbols);
```

**Review by esanuandra — CHANGES_REQUESTED** [review]
In `helpers.js`, function `createGraphData` add the `queued` flag

```
export const createGraphData = (
...
) =>
  seriesData.map((series) => {
    ...
    return {
      color: color || ['border-secondary', ''],
      symbol: symbol || ['circle', 'outline'],
      visible: Boolean(color),
      // true only when the palette was exhausted at creation (auto-queued/
      // force-hidden). removeTest promotes these; user-hidden tests stay hidden.
      queued: !color,
      name: series.name,
      suite: series.suite,
      signature_id: series.signature_id,
      signatureHash: series.signature_hash,
```

**Review by esanuandra — APPROVED** [review]
lgtm
